src/GraphQL/Types/Collections/TypeFields.cs

```diff
@@ -0,0 +1,37 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace GraphQL.Types
+{
+    /// <summary>
+    /// A class that represents a set of fields for <see cref="IComplexGraphType"/> i.e <see cref="ComplexGraphType{TSourceType}"/>.
+    /// </summary>
+    public class TypeFields : IEnumerable<FieldType>
+    {
+        internal List<FieldType> List { get; } = new List<FieldType>();
+
+        /// <summary>
+        /// Gets the count of fields.
+        /// </summary>
+        public int Count => List.Count;
+
+        internal void Add(FieldType field)
+        {
+            if (field == null)
+                throw new ArgumentNullException(nameof(field));
+
+            if (!List.Contains(field))
+                List.Add(field);
+        }
+
+        public bool Contains(FieldType field) => List.Contains(field ?? throw new ArgumentNullException(nameof(field)));
+
+        public bool Contains(IFieldType field) => List.Contains((FieldType)field ?? throw new ArgumentNullException(nameof(field)));
+
+        /// <inheritdoc cref="IEnumerable.GetEnumerator"/>
+        public IEnumerator<FieldType> GetEnumerator() => List.GetEnumerator();
```

Comment:
I took a look in more detail.  It seems that this idea might work well for some of the new classes, and not so much for others.  Your call if you want to implement it or not.

Reply:
> so we may (or may not) be able to remove the internal List<FieldType> List property

Why delete it? Moreover `List<FieldType>.Enumerator` exposes implementation details. All these classes are needed by us rather than by the library clients. It is unlikely that anyone uses these APIs often. I deliberately left the compatibility for `IEnumerable<>` with external code but all internal GraphQL.NET bits began to use internal properties (List).

Reply:
Well, we could write our own struct enumerator, but the one provided by `List` works just fine.  I don't really care, but I thought this would provide a cleaner codebase for us to work within, as we can use a standard `foreach` syntax without having to access `List`.  And less chance of error within future changes.  Moreover, anyone else who may ever enumerate through arguments or fields on a graph will also benefit. It's still fully `IEnumerator` compatible, and the members are public bits of .NET Standard 2.0.  I'm a bit surprised you don't like the idea, that's all.  It doesn't make much difference to me.

Reply:
> `List<FieldType>.Enumerator` exposes implementation details. 

Yeah, if we did implement it, it would be hard to change the internals to anything except a list.  And writing custom struct enumerators for each class isn't worth it.

Reply:
Yes. It much easier to type `.List`.